src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2103,8 +2103,60 @@ cdef class Polynomial(CommutativePolynomial):
             sage: r = f.any_root()
             sage: r^2 + r
             a^2 + a
-        """
-        if self.base_ring().is_finite() and self.base_ring().is_field():
+
+        Check that issue 33329 is resolved: implementation used to loop infinitely
+
+            sage: k = GF(2^60)
+            sage: R.<x> = k[]
+            sage: a, b = k.random_element(), k.random_element()
+            sage: f = (x + a) * (x + b)
+            sage: f.any_root() in (a, b)
+            True
+            sage: f = (x + a) * (x + b + 1)
+            sage: f.any_root() in (a, b + 1)
+            True
+        """
+        # For finite fields, perform partial factorization until an irreducible
+        # polynomial of suitable degree is found.
+        # Reference: https://www.shoup.net/papers/frobenius.pdf
+        def qfrobenius(modulus, k, xq):
+            """
+            Compute the q=p^k power Frobenius F in quotient ring K[x]/modulus
+            """
+            # If pol = sum(pol[i] x^i), F(pol) = sum(F(pol[i]) xq^i)
+            # which is a modular evaluation. We use Brent-Kung method:
+            # Writing indices i = at+b the sum becomes:
+            # sum(a=0..deg/t, xq^at * sum(b=0..t, F(pol[at+b]) xq^b))
+            p = modulus.parent().characteristic()
+            xq %= modulus
+            if k == 1 and p < 100:
+                return lambda pol: pow(pol, p, modulus)
+            t = modulus.degree().isqrt() + 1
+            smalls = [1, xq]
+            while len(smalls) < t:
+                smalls.append((smalls[-1] * xq) % modulus)
+            xqt = (smalls[-1] * xq) % modulus
+            if modulus.parent().base_ring().degree() == k:
+                scalarfrob = lambda x: x
+            else:
+                scalarfrob = lambda x: x.frobenius(k)
+
+            def f(pol):
+                if pol == 0:
+                    return pol
+                blocks = [
+                    sum(scalarfrob(pol[a+b])*smalls[b] for b in range(t) if a+b <= pol.degree())
+                    for a in range(0, pol.degree()+1, t)
+                ]
```

Comment:
It is generally (always?) better to unroll these as `for` loops in Cython, also to initialize a list (which you could `cdef list blocks` too) to something like `blocks = [None]*(pol.degree()//t)`. Also, avoid the `range(x,y,t)` since Cython does not optimize that. Instead manually do the step as a multiplication, e.g., `for a in range(x,y//t)` with `a*t` in the code.